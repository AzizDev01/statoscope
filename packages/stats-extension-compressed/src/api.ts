import {
  default as makeEntityResolver,
  Resolver,
} from '@statoscope/helpers/dist/entity-resolver';
import { APIFactory } from '@statoscope/extensions';
import { Format, Resource, Size } from './generator';

export type API = (compilationId: string, resourceId: string) => Size | null;

const makeAPI: APIFactory<Format, API> = (source) => {
  const sizeResolvers: Map<string, Resolver<string, Resource>> = new Map();

  for (const compilation of source.payload.compilations) {
    sizeResolvers.set(
      compilation.id,
      makeEntityResolver(compilation.resources, (r) => r.id)
    );
  }

  return (compilationId: string, resourceId: string): Size | null => {
    return sizeResolvers.get(compilationId)?.(resourceId)?.size ?? null;
  };
};

export default makeAPI;
