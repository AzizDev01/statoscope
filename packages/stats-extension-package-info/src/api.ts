import {
  default as makeEntityResolver,
  Resolver,
} from '@statoscope/helpers/dist/entity-resolver';
import { APIFactory } from '@statoscope/extensions';
import { Format, Package, Instance } from './generator';

export type API = {
  getPackage: (compilationId: string, packageName: string) => Package | null;
  getInstance: (
    compilationId: string,
    packageName: string,
    instancePath: string
  ) => Instance | null;
};

const makeAPI: APIFactory<Format, API> = (source) => {
  const packageResolvers: Map<string, Resolver<string, Package>> = new Map();
  const instanceResolvers: Map<Package, Resolver<string, Instance>> = new Map();

  for (const compilation of source.payload.compilations) {
    packageResolvers.set(
      compilation.id,
      makeEntityResolver(compilation.packages, (r) => r.name)
    );

    for (const packageItem of compilation.packages) {
      instanceResolvers.set(
        packageItem,
        makeEntityResolver(packageItem.instances, (r) => r.path)
      );
    }
  }

  return {
    getPackage(compilationId: string, packageId: string): Package | null {
      return packageResolvers.get(compilationId)?.(packageId) ?? null;
    },
    getInstance(
      compilationId: string,
      packageId: string,
      instancePath: string
    ): Instance | null {
      const resolvedPackage = packageResolvers.get(compilationId)?.(packageId) ?? null;

      if (!resolvedPackage) {
        return null;
      }

      return instanceResolvers.get(resolvedPackage)?.(instancePath) ?? null;
    },
  };
};

export default makeAPI;
