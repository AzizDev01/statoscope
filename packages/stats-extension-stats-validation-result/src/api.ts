import makeEntityResolver from '@statoscope/helpers/dist/entity-resolver';
import { Resolver } from '@statoscope/helpers/dist/entity-resolver';
import { APIFactory } from '@statoscope/extensions';
import { RelatedItem, RuleDescriptor } from '@statoscope/types/types/validation';
import { Format, Item } from './generator';

export type API = {
  getItems: (
    compilationId: string | null,
    type?: RelatedItem['type'] | null,
    relatedId?: string | number
  ) => Item[];
  getItemById(id: number): Item | null;
  getRule(id: string): RuleDescriptor | null;
};

type IndexItem = { item: Item; related: RelatedItem };
type Index = {
  type: RelatedItem['type'];
  links: IndexItem[];
  items: Item[];
  related: RelatedItem[];
  byRelatedId: Resolver<string | number, Item>;
};

export type IndexAPI = {
  readonly type: RelatedItem['type'];
  add(item: Item, related: RelatedItem): void;
  getItems(id?: string | number): Item[];
};

function makeIndex(type: RelatedItem['type']): IndexAPI {
  const links: IndexItem[] = [];
  const index: Index = {
    type,
    links,
    items: [],
    related: [],
    byRelatedId: makeEntityResolver(
      links,
      (item) => item.related.id,
      (item) => item.item
    ),
  };

  return {
    type,
    add(item: Item, related: RelatedItem): void {
      if (!index.items.includes(item)) {
        index.items.push(item);
      }

      if (!index.related.includes(related)) {
        index.related.push(related);
      }

      if (!index.links.find((link) => link.item === item && link.related === related)) {
        index.links.push({ item, related });
      }
    },
    getItems(id?: string | number): Item[] {
      if (id != null) {
        return (
          index.links
            // eslint-disable-next-line eqeqeq
            .filter((link) => link.related.id == id)
            .map((link) => link.item)
        );
      }

      return index.items;
    },
  };
}

export type CompilationData = {
  id: string | null;
  items: Item[];
  resolveIndex: Resolver<RelatedItem['type'], IndexAPI>;
};

const makeAPI: APIFactory<Format, API> = (source) => {
  const compilationResolvers: CompilationData[] = [];
  const resolveCompilationsResolvers = makeEntityResolver(
    compilationResolvers,
    (item) => item.id
  );
  const resolveRule = makeEntityResolver(
    source.payload.rules,
    (item) => item.name,
    (item) => item.descriptor
  );
  const items: Item[] = [];
  const resolveItemById = makeEntityResolver(items, (item) => item.id);
  for (const compilation of source.payload.compilations) {
    const indexes: IndexAPI[] = [];
    const compilationData: CompilationData = {
      id: compilation.id,
      items: compilation.items,
      resolveIndex: makeEntityResolver(indexes, (index) => index.type),
    };
    compilationResolvers.push(compilationData);

    for (const item of compilation.items) {
      items.push(item);
      for (const related of item.related) {
        let index = compilationData.resolveIndex(related.type);

        if (!index) {
          index = makeIndex(related.type);
          indexes.push(index);
        }

        index.add(item, related);
      }
    }
  }

  return {
    getItems: (
      compilationId: string | null,
      type?: RelatedItem['type'] | null,
      relatedId?: string | number
    ): Item[] => {
      if (relatedId) {
        if (!type) {
          throw new Error('type must be specified');
        }

        return (
          resolveCompilationsResolvers(compilationId)
            ?.resolveIndex(type)
            ?.getItems(relatedId) ?? []
        );
      }

      if (type) {
        return (
          resolveCompilationsResolvers(compilationId)?.resolveIndex(type)?.getItems() ??
          []
        );
      }

      return resolveCompilationsResolvers(compilationId)?.items ?? [];
    },
    getRule(id: string): RuleDescriptor | null {
      return resolveRule(id);
    },
    getItemById(id: number): Item | null {
      return resolveItemById(id);
    },
  };
};

export default makeAPI;
