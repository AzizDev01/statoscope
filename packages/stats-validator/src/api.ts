import { API, Storage, RuleDescriptor } from '@statoscope/types/types/validation/api';

export function makeAPI(): API {
  let descriptor: RuleDescriptor | null = null;
  const storage: Storage = [];
  return {
    setRuleDescriptor(desc: RuleDescriptor): void {
      descriptor = desc;
    },
    getRuleDescriptor(): RuleDescriptor | null {
      return descriptor;
    },
    getStorage(): Storage {
      return storage;
    },
    message(text, options): void {
      storage.push({
        filename: options?.filename,
        compilation: options?.compilation,
        details: options?.details,
        related: options?.related,
        message: text,
      });
    },
  };
}
