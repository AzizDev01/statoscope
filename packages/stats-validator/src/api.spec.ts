import { makeAPI } from './api';

test('error', () => {
  const api = makeAPI();

  api.message('foo-message');
  api.message('bar-message', { filename: 'foo-filename' });
  api.message('baz-message', {
    filename: 'foo-filename',
    compilation: 'foo-compilation',
    details: [{ type: 'text', content: 'foo' }],
    related: [
      { id: 'foo-module', type: 'module' },
      { id: 'foo-package', type: 'package' },
    ],
  });

  api.setRuleDescriptor({
    description: ' foo',
    package: { name: 'foo', version: '1.0.0' },
  });
  expect(api.getRuleDescriptor()).toMatchSnapshot();
  expect(api.getStorage()).toMatchSnapshot();
});
