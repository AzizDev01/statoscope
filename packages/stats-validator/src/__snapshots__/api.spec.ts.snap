// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error 1`] = `
Object {
  "description": " foo",
  "package": Object {
    "name": "foo",
    "version": "1.0.0",
  },
}
`;

exports[`error 2`] = `
Array [
  Object {
    "compilation": undefined,
    "details": undefined,
    "filename": undefined,
    "message": "foo-message",
    "related": undefined,
  },
  Object {
    "compilation": undefined,
    "details": undefined,
    "filename": "foo-filename",
    "message": "bar-message",
    "related": undefined,
  },
  Object {
    "compilation": "foo-compilation",
    "details": Array [
      Object {
        "content": "foo",
        "type": "text",
      },
    ],
    "filename": "foo-filename",
    "message": "baz-message",
    "related": Array [
      Object {
        "id": "foo-module",
        "type": "module",
      },
      Object {
        "id": "foo-package",
        "type": "package",
      },
    ],
  },
]
`;
